propolis/src/dispatch/async_tasks.rs

```diff
@@ -0,0 +1,168 @@
+use std::collections::hash_map::HashMap;
+use std::sync::{Arc, Mutex};
+
+use futures::future::BoxFuture;
+use tokio::runtime::{Builder, Runtime};
+use tokio::sync::Semaphore;
+use tokio::task::JoinHandle;
+
+use super::{DispCtx, SharedCtx};
+
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+pub struct AsyncTaskId {
+    val: usize,
+}
+
+struct AsyncTask {
+    hdl: JoinHandle<()>,
+    ctx_sem: Arc<Semaphore>,
+    task_sem: Arc<Semaphore>,
+    quiesced: bool,
+}
+
+#[derive(Default)]
+struct AsyncInner {
+    next_id: usize,
+    quiesced: bool,
+    shutdown: bool,
+    tasks: HashMap<AsyncTaskId, AsyncTask>,
+}
+impl AsyncInner {
+    fn next_id(&mut self) -> AsyncTaskId {
+        self.next_id += 1;
+        AsyncTaskId { val: self.next_id }
+    }
+}
+
+pub(super) struct AsyncDispatch {
+    inner: Mutex<AsyncInner>,
+    rt: Mutex<Option<Runtime>>,
```

Comment:
1. Why let this be optional? It seems like it should always exist, if I understand the constructor usage correctly
2. Would that avoid the need for the `Mutex`? Most `Runtime` methods act on `&self`, not `&mut self`.

Reply:
I ran into issues where the Runtime was dropped from an async context in propolis-server (When the `Instance` -> `Dispatcher` -> `Runtime` drop chain occurred).  That's why it's in an option so I can drop it separately from a known sync context.